pkg/kubelet/nodelease/controller.go

```diff
@@ -0,0 +1,176 @@
+	clientset "k8s.io/client-go/kubernetes"
+	coordclientset "k8s.io/client-go/kubernetes/typed/coordination/v1beta1"
+	"k8s.io/kubernetes/pkg/util/pointer"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// renewFactor is the fraction of the lease duration that the Kubelet waits before renewing the lease
+	// TODO(mtaufen): determine a good renewFactor, this should be a standard behavior and should not need to be configured
+	renewFactor = 0.33
+	// maxUpdateRetries is the number of immediate, successive retries the Kubelet will attempt
+	// when renewing the lease before it waits for the renewal interval before trying again,
+	// similar to what we do for node status retries
+	maxUpdateRetries = 5
+)
+
+// Controller manages creating and renewing the lease for this Kubelet
+type Controller interface {
+	Run(stopCh <-chan struct{})
+}
+
+var _ Controller = &controller{}
+
+type controller struct {
+	client                     coordclientset.LeaseInterface
+	holderIdentity             string
+	leaseDurationSeconds       int32
+	renewInterval              time.Duration
+	clock                      clock.Clock
+	onRepeatedHeartbeatFailure func()
+}
+
+// NewController constructs and returns a controller
+func NewController(clock clock.Clock, client clientset.Interface, holderIdentity string, leaseDurationSeconds int32, onRepeatedHeartbeatFailure func()) Controller {
+	var leaseClient coordclientset.LeaseInterface
+	if client != nil {
+		leaseClient = client.CoordinationV1beta1().Leases(metav1.NamespaceNodeLease)
+	}
+	return &controller{
+		client:               leaseClient,
+		holderIdentity:       holderIdentity,
+		leaseDurationSeconds: leaseDurationSeconds,
+		renewInterval:        time.Duration(float64(time.Duration(leaseDurationSeconds)*time.Second) * renewFactor),
+		clock:                clock,
+		onRepeatedHeartbeatFailure: onRepeatedHeartbeatFailure,
+	}
+}
+
+// Run runs the controller
+func (c *controller) Run(stopCh <-chan struct{}) {
+	if c.client == nil {
+		glog.Infof("node lease controller has nil client, will not claim or renew leases")
+		return
+	}
+	wait.Until(c.sync, c.renewInterval, stopCh)
+}
+
+func (c *controller) sync() {
+	lease, created := c.backoffEnsureLease()
+	// we don't need to update the lease if we just created it
+	if !created {
+		c.retryUpdateLease(lease)
+	}
+}
+
+// backoffEnsureLease attempts to create the lease if it does not exist,
+// and uses exponentially increasing waits to prevent overloading the API server
+// with retries. Returns the lease, and true if this call created the lease,
+// false otherwise.
+func (c *controller) backoffEnsureLease() (*coordv1beta1.Lease, bool) {
+	var (
+		lease   *coordv1beta1.Lease
+		created bool
+		err     error
+	)
+	sleep := 100 * time.Millisecond
+	for {
+		lease, created, err = c.ensureLease()
+		if err == nil {
+			break
+		}
+		// log error
+		glog.Errorf("failed to ensure node lease exists, error: %v", err)
+		// backoff wait
+		time.Sleep(sleep)
```

Comment:
is there a reason we're not using the injected clock here?

Reply:
oops, good catch!

Reply:
done